src/packages/version-check/src/version-check.ts

```diff
@@ -0,0 +1,321 @@
+
+      this._session = session;
+      this._request = req;
+
+      session.on("error", err => reject(err));
+      req.setEncoding("utf8");
+
+      req
+        .on("error", reject)
+        .on("response", (_headers, _flags) => {
+          let data: string = "";
+          req
+            .on("data", chunk => {
+              data += chunk;
+            })
+            .on("end", async () => {
+              resolve(data);
+              session.close();
+            });
+        })
+        .setTimeout(ttl, () => {
+          this.destroy();
+          reject(`ttl expired: ${ttl}`);
+        });
+      req.end();
+    });
+  }
+
+  /**
+   * Shows the banner message to the user on ganache startup.
+   *
+   */
+  log() {
+    // We intentionally do not wait for this. We are always comparing against the
+    // previously fetched version.
+    this.getLatestVersion();
+    if (this.canNotifyUser()) {
+      const currentVersion = this._currentVersion;
+      const { packageName, latestVersion } = this._config;
+
+      // squelch any errors - semver can throw, _logger.log can be user-defined
+      // and _updateConfig writes to fs.
+      try {
+        const upgradeType = semverUpgradeType(currentVersion, latestVersion);
+
+        this._logger.log(
+          bannerMessage({
+            upgradeType,
+            packageName,
+            currentVersion,
+            latestVersion
+          })
+        );
+
+        this._updateConfig({
+          lastNotification: Date.now(),
+          lastVersionLogged: latestVersion
+        });
+      } catch {}
+    }
+  }
+
+  /**
+   * Appended to `ganache --version` if a newer version is known. This always runs
+   * regardless of whether enabled/disabled.
+   *
+   * In DEV mode (e.g. version === 'DEV') this._currentVersion will be undefined because
+   * 'DEV' is invalid semver and VC will disable in the constructor. Even though cliMessage
+   * runs regardless of enabled/disabled it is not possible to calculate an upgrade type for 'DEV'
+   * and semverUpgradeType will throw on undefined.
+   */
+  cliMessage() {
+    if (!this._currentVersion) return "";
+
+    const currentVersion = this._currentVersion;
+    const { latestVersion } = this._config;
+
+    // squelch any errors - semver package can throw though it is unlikely to here.
+    try {
+      if (semverUpgradeType(currentVersion, latestVersion)) {
+        return `note: there is a new version available! ${currentVersion} -> ${latestVersion}`;
+      }
+    } catch {}
+    return "";
+  }
+
+  static get DEFAULTS(): VersionCheckOptions {
+    return {
+      packageName: "ganache",
+      enabled: true,
+      url: "https://version.trufflesuite.com",
+      ttl: 30000, // http2session.setTimeout
+      latestVersion: "0.0.0", // Last version fetched from the server
+      lastVersionLogged: "0.0.0", // Last version to tell the user about
+      lastNotification: 0,
+      disableInCI: true,
+      activated: false
+    };
+  }
```

Comment:
why?

Reply:
TLDR Yes, but really

Selfishly, the idiom makes me gruntled. I think of a config as a class property, not a data field or function method (to your point). This plays well with other config sources, is a nice defensive fallback to anything exposed to a user, and I started doing it before I got to TS and `import/export` 

I wasn't sure if the suggestion was:

`export const config = {...}` or adding `private _config: {...}` or making each config property it's own property on the class, storing it in a CONST file, etc,  was the suggestion (or something else I didn't consider)

Reply:
Ok, I tried to ironman `const DEFAULTS = {...}` because I figured functionally it was probably equivalent and wouldn't create a new object every time. 

So I was like, @davidmurdoch "But isn't a `const` encapsulated so how would it be accessed` and then he was like "What's this dark magic" and I was like, oh dang. 

So if that was the suggestion I believe the tldr is: You could access the `const` on the prototype, make changes, and sham bam bamina.

If that was not what you meant, sorry!